pkg/idp/idp.go

```diff
@@ -96,34 +96,31 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	}
 	{
 		// Get existing connectors from the dex config secret
-		oldConnectors, err := getConnectorsFromSecret(secret)
+		oldConfig, err := getDexConfigFromSecret(secret)
 		if err != nil {
 			return microerror.Mask(err)
 		}
+		oldConnectors := getConnectorsFromConfig(oldConfig)
 		// Create apps for each provider and get dex config
 		appConfig, err := s.GetAppConfig(ctx)
 		if err != nil {
 			return microerror.Mask(err)
 		}
-		dexConfig, err := s.CreateOrUpdateProviderApps(appConfig, ctx, oldConnectors)
+		newConfig, err := s.CreateOrUpdateProviderApps(appConfig, ctx, oldConnectors)
 		if err != nil {
 			return microerror.Mask(err)
 		}
-		data, err := json.Marshal(dexConfig)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-		// Fetching the newest version of the secret. If we are here we assume that it exists.
-		if err := s.Get(ctx, types.NamespacedName{Name: dexSecretConfig.Name, Namespace: dexSecretConfig.Namespace}, secret); err != nil {
-			return microerror.Mask(err)
-		}
-		secret.Data = map[string][]byte{"default": data}
-		// Get new connectors from the dex config secret
-		newConnectors, err := getConnectorsFromSecret(secret)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-		if updateSecret := s.secretNeedsUpdate(oldConnectors, newConnectors); updateSecret {
+
+		if updateSecret := s.secretDataNeedsUpdate(oldConfig, newConfig); updateSecret {
+			data, err := json.Marshal(newConfig)
```

Comment:
We can serialize the config and apply it to the latest secret only in case there is a need for update.